examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,227 @@
+// -*- mode: Bluespec; -*-
+
+/**
+ * A specification ICS20 fungible token transfer protocol.
+ *
+ * For now, this only models the denomination trace changes between transfers.
+ *
+ * Gabriela Moreira, Informal Systems, 2023
+ */
+module denomTrace {
+  type Hop = {
+    port: str,
+    channel: str,
+  }
+
+  type DenomTrace = {
+    path: List[Hop],
+    baseDenom: str,
+  }
+
+  pure def senderIsSource(sourcePort: str, sourceChannel: str, denomTrace: DenomTrace): bool =
+    denomTrace.path.length() >= 1 and
+    denomTrace.path.head().port == sourcePort and
+    denomTrace.path.head().channel == sourceChannel
+
+  pure def receiverIsSource(sourcePort: str, sourceChannel: str, denomTrace: DenomTrace): bool =
+    not(senderIsSource(sourcePort, sourceChannel, denomTrace))
+
+  pure def updateTrace(denomTrace: DenomTrace, sourcePort: str, sourceChannel: str, destPort: str, destChannel: str): DenomTrace = {
+    if (senderIsSource(sourcePort, sourceChannel, denomTrace)) {
```

Comment:
Should not this be `receiverIsSource` actually?

I still find the naming in ICS20 quite confusing. I know that this is how it called in the IBC specs. However, could we find better names, and say in comments, how these things are called in ICSes?

The confusing thing here is when we check `senderIsSource`, we are not talking about the same "source" as in `sourceChannel` and `sourcePort`. In this context, the source is the chain that has sent the coins in the past, and now we are traversing back.

Would not it make sense to rename "source" in something like "origin"? The point is that a denomination was produced on a path from the origin at some point. On every step, we are either: (1) Getting closer to the origin, or (2) Getting further away from the origin. The source and destination are then reserved for the packet that is being processed.

Reply:
Makes sense to me! Good call :smile: